[#49782] Fix out-of-bounds widgets error when creating Version Boards

aaron-contreras opened the pull request:
## Bug description
An ["out-of-bounds" validation](https://github.com/opf/openproject/blob/21a696ef9b170e14ad2daf53364a4c2113822c2f/modules/grids/app/contracts/grids/base_contract.rb#L139-L144) was rightfully raising an error since when a Project had more than 4 versions which applied for being a column on a version board, the amount of columns being set on the board was 4 (as done in the original implementation of Board creation).

Version boards however, are a special case, as the number of columns to be created initially with it is varying based on the amount of open versions for the project at the moment of creating the board.

## Fix
The `VersionBoardCreateService` now sets the `column_count_for_board` dynamically based on the versions that a widget will be created for; this ensures they're mapped 1-to-1 (or the default if no versions apply) and we don't run into out-of-bounds widgets for them anymore.

Added a modification to the setup code in the `VersionBoardCreateService` spec file demonstrates the failing behavior and success behavior after the override.
## Notes
See Bug Work Package: https://community.openproject.org/work_packages/49782

aaron-contreras committed:
Modify spec to create a failing number of versions

Any amount of versions over 4 in a project would fail validation by
the `VersionBoardCreateService` due to a widget being out of bounds.

This is due to the default number of columns created for a Grid is
set to 4 (as in legacy implementation).

aaron-contreras committed:
Override `column_count_for_board`

Sets the number of columns dynamically based on the amount of widgets
that are going to be created; in this case: the number of versions that
match the criteria for being a column in the board.

cbliard reviewed (changes requested):
There is a failing test, and it highlights an issue when the project has no versions:

![image](https://github.com/opf/openproject/assets/176055/d3256dd2-a198-4bf6-b3e2-ca2b9dcb0650)

failing test is `modules/boards/spec/features/boards_sorting_spec.rb:49`

aaron-contreras committed:
Ensure `column_count` is the default if no versions apply

Sets the column_count to the default for grids (4), if no widgets
are to be created, preventing the minimum column count validation
error from being raised and behaving like other grids if the special
case does not apply.

aaron-contreras commented:
Thanks for catching that bug @cbliard !

cbliard commented:
> Thanks for catching that bug @cbliard !

Let's be honest, I did not catch anything: it's an existing test which detected the issue. The person to thank is @bsatarnejad. Thank you!
